NDTensors/src/dense/dense.jl

```diff
@@ -20,38 +20,43 @@ struct Dense{ElT,VecT<:AbstractArray} <: TensorStorage{ElT}
   end
 end
 
-function Dense(data::VecT) where {VecT<:AbstractArray{ElT}} where {ElT}
-  return Dense{ElT,VecT}(data)
+## We need a general typeinfo constructor for Dense
+default_Densetype(datatype::Type{<:AbstractArray{ElT}}) where {ElT} = default_storagetype(datatype)
+default_Densetype(eltype::Type{<:Number}) = default_Densetype(default_datatype(eltype))
+default_Densetype() = default_Densetype(default_datatype(default_eltype()))
+## End typeinfo based constructors 
+
+#Start with high information constructors and move to low information constructors
+function Dense{ElT, VecT}() where {ElT, VecT<:AbstractArray{ElT}}
+  return default_Densetype(VecT)(VecT())
 end
 
-function Dense(data::Array{ElT}) where {ElT}
-  return Dense{ElT,Vector{ElT}}(vec(data))
+# Construct from a set of indices
+function Dense{ElT,VecT}(inds::Tuple) where {ElT,VecT<:AbstractArray{ElT}}
+  return default_Densetype(VecT)(VecT(undef, dim(inds)))
+  #eturn Dense(VecT(undef, dim(inds)))
 end
 
+# This function is ill-defined. It cannot transform a complex type to real...
 function Dense{ElR}(data::AbstractArray{ElT}) where {ElR,ElT}
-  return ElT == ElR ? Dense(data) : Dense(ElR.(data))
+  return default_Densetype(ElR)(data)
 end
 
-# Construct from a set of indices
-function Dense{ElT,VecT}(inds) where {ElT,VecT<:AbstractArray{ElT}}
-  return Dense(VecT(dim(inds)))
-end
+#Do we want this to be zero? 
+Dense{ElT}(dim::Integer) where {ElT} = default_Densetype(ElT)(default_datatype(ElT)(undef, dim))
```

Comment:
I don't think `default_Densetype(ElT)` is needed here, it seems sufficient to just use a `Dense` constructor directly.

By default I think this should make a storage filled with zeros, and only make `undef` if `undef` is specified. But that touches on a question about how to make a generic data that is filled with zeros.

We may need to define additional generic functions for that like:
```julia
zeros(datatype::Type{<:AbstractArray}, dims) = fill!(datatype(undef, dims), zero(eltype(datatype)))
```